ta/pkcs11/src/pkcs11_token.c

```diff
@@ -973,11 +973,57 @@ static enum pkcs11_rc set_pin(struct pkcs11_session *session,
 		return PKCS11_CKR_SESSION_READ_ONLY;
 
 	if (IS_ENABLED(CFG_PKCS11_TA_AUTH_TEE_IDENTITY) &&
-	    token->db_main->flags & PKCS11_CKFT_PROTECTED_AUTHENTICATION_PATH) {
-		rc = setup_identity_auth_from_pin(token, user_type, new_pin,
-						  new_pin_size);
-		if (rc)
+	    (token->db_main->flags &
+	     PKCS11_CKFT_PROTECTED_AUTHENTICATION_PATH)) {
+		rc = setup_identity_auth_from_pin(token, user_type,
+						  new_pin, new_pin_size);
+		if (rc == PKCS11_CKR_OK) {
+			goto update_db;
+		} else if (rc == PKCS11_CKR_PIN_INVALID &&
+			   !(token->db_main->flags &
+		      PKCS11_CKFT_USER_PIN_INITIALIZED)) {
+			/*
+			 * PIN was not compatible with TEE Identity
+			 * Authentication syntax so let's assume it might be a
+			 * new SO PIN to switch the authentication mode.
+			 */
+
+			/* Update mode flag if all PIN checks pass */
+			flags_clear |=
+				PKCS11_CKFT_PROTECTED_AUTHENTICATION_PATH;
+		} else {
+			return rc;
+		}
+	} else if ((user_type == PKCS11_CKU_SO) && !new_pin &&
+		   !(token->db_main->flags &
+		     PKCS11_CKFT_USER_PIN_INITIALIZED)) {
+		/*
+		 * Allow changing of token authentication mode before user pin
+		 * has been initialized.
```

Comment:
Why does it matter? How to change once the user PIN has been set? The SO should be able to do anything I think.

Reply:
Problem here is that it is not feasible to mix authentication modes for SO and USER. You need to change mode synchronously for both.

This problem comes from how storage for PKCS#11 is handled and that is a bit hard to change as this feature is now published and in the field.

Authentication mode is controlled by following flag:
```c
token->db_main->flags & PKCS11_CKFT_PROTECTED_AUTHENTICATION_PATH
```

PKCS#11 specification 2.40 does not understand other users than SO and USER. Next version 3.0+ does provide some idea that there could be more users.

Now users are really stored in storage as two PINs (one for SO and one for USER) (`struct token_persistent_main`). In here PIN can include both numbers and characters (and other glyphs).

TEE Identity is shared storage with PIN so whatever authentication mode you have configured credentials cannot be changed independently of the SO PIN and USER PIN values stored in storage as that could give very weird results.

Only ways to do that in controlled way is to do that in token initialization time or when the USER PIN is not set as then we have only one PIN to change.

In token initialization time we can only do once at the beginning as in successive cases specification states that you need to authenticate with existing credential. (and this is related to the problem in the previous code).

Now if we want to switch the authentication mode we cannot use the token initialization point with the standard APIs.

Now the next spot is the PIN change.

If the user PIN has not been set then we do not need to care about its value so we can figure out if we have some method that we could change the mode in here.

If we are in PIN authentication mode then we allow mode change by if the new pin is empty. In here we copy the client's TEE Identity for the connection to the credential and then change the authentication mode. To change to other TEE Identity one need to issue another PIN change so then one can use more fine grained control. This limitation is due to fact that you can have textual PIN also so in theory there could be clash -- I believe this is OK limitation as you should not be changing the authentication mode in normal operation environment.

If we are in TEE identity authentication mode then we allow mode change by giving non-conforming TEE identity string. In theory in here we also have issue of possible clash but then you need to just go for temporary PIN during mode change and then go for final PIN after the authentication mode change.

Last but not least -- 
- If USER would have been allowed to change the authentication mode then it could lock out the SO user completely as then the PIN would be garbage. 
- If SO would have changed the authentication mode while valid USER exists then the USER pin would be garbage.

Reply:
Thanks for the detailed explanation.

> If SO would have changed the authentication mode while valid USER exists then the USER pin would be garbage.

In this case, could we simply clear the user PIN maybe?

Reply:
Kinda feels wrong thing to do but @etienne-lms comments on this?

Reply:
User PIN is cleared when token is initialized.
So switching auth mode should work right after a call to `C_TokenInit()`.